woocommerce/woocommerce-ios
# Edit Grouped Products: edit linked products (Part 3/3) - search product list selector

Search product list selector for grouped product's linked products #2199

This PR implements the highlighted screen in the flow to search and select linked products for a grouped product:

![search-linked-products](https://user-images.githubusercontent.com/1945542/88175270-3a912300-cc58-11ea-893d-7a43bce8e35b.png)

## Changes

Note: I understand that there are a bunch of changes to make the UI sync with the selected product IDs from multiple sources (search UI and product list selector). Happy to explain any changes further, and feel free to suggest a better way for these UI components to interact with each other!

- In `PaginatedListSelectorViewController` (the child view controller of the screen that presents search UI), added `reloadData` function to allow the paginated list selector to reload data in case of data changes not from selecting from the list
- In `ProductListMultiSelectorDataSource` (the data source for the screen that presents search UI), added `addProducts` function to update the selected product IDs externally
- Changes in `SearchUICommand` protocol:
  - Added `func configureActionButton(_ button: UIButton, onDismiss: @escaping () -> Void)` so that the search UI can display different text on the dismiss CTA ("Done" when there are any selected products, and "Cancel" otherwise) and handle the selected product IDs. Before this PR and use case, the dismiss CTA always shows "Cancel".
  - Added `reloadData: () -> Void` and `updateActionButton: () -> Void` to `func didSelectSearchResult` so that the search UI can reload the search results to reflect the selected/unselected UI change and trigger an update on the action button configuration for different button text via `configureActionButton` above.
- Updated existing `SearchUICommand` implementations (`OrderSearchUICommand` and `ProductSearchUICommand`) based on the protocol changes
- Created `ProductListMultiSelectorSearchUICommand` that implements `SearchUICommand` for selecting multiple products from search UI
- In `ProductListSelectorViewController`, presented `SearchViewController` with `ProductListMultiSelectorSearchUICommand`, updated its data source on any selected product IDs from search and then reloaded the paginated product list
- In `SearchViewController`, updated based on changes in `SearchUICommand` protocol. Additionally, it now keeps track of the command's results controller's `predicate` to be used along with the search query predicate. This is because this product search UI selector has a predicate to exclude certain product IDs, and the view controller used to remove the pre-existing predicate on search query changes.
  - This also fixes an issue in existing product search UI, where the site predicate gets removed whenever the search query changes. I just noticed this bug after switching sites.
- In Yosemite/Networking layers, added `excludedProductIDs` param to `searchProducts`

## Testing

- Go to the Products tab
- Tap on a grouped product
- Tap on the "Grouped products" row
- Tap on the "Add Products" button 
- Tap on the search icon in the navigation bar --> should see a list of products that **do not contain the pre-selected products plus the grouped product itself** (even though it's allowed to select itself in core)
- Select/unselect some products --> the "Cancel" button's text should change to "Done" when at least one product is selected 
- Search for some products --> when there are no results, a placeholder should be shown
- Select some products from the search results or initial screen where the search query is empty
- Tap "Done" --> the selected products should show as selected in the "Grouped Products" list selector with the number of selected products in the navigation bar
- Tap "Done" --> the "Grouped products" row should show the updated total number of linked products
- Tap "Update" --> the linked products for the grouped product should be updated remotely

### Sanity check on affected screens

- Go to the Products tab
- Tap on the search icon and make sure the search works as before
- Go to the Orders tab
- Tap on the search icon and make sure the search works as before

## Example screenshots

![simulator](https://user-images.githubusercontent.com/1945542/88179386-67483900-cc5e-11ea-8871-eeef9b27dc1b.gif)

\ | No results selected | Non-zero results selected
-- | -- | --
Light | ![Simulator Screen Shot - iPhone 11 - 2020-07-22 at 21 00 03](https://user-images.githubusercontent.com/1945542/88179369-60b9c180-cc5e-11ea-86b4-4ba14de01d94.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-22 at 21 00 05](https://user-images.githubusercontent.com/1945542/88179377-657e7580-cc5e-11ea-9f77-1fe01946458e.png)
Dark | ![Simulator Screen Shot - iPhone 11 - 2020-07-22 at 21 00 21](https://user-images.githubusercontent.com/1945542/88179383-66170c00-cc5e-11ea-800d-1b9d529c1f87.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-22 at 21 00 22](https://user-images.githubusercontent.com/1945542/88179385-66afa280-cc5e-11ea-92ec-23b22fe7a7e3.png)

Update release notes:
